### ISSUE TITLE
Compilation failed under Android 25

### ISSUE THREAD

**Issue by jimwang118** [randombit/botan#3890]
When I compile botan under Android, the following error occurs.
```
src/lib/utils/prefetch.cpp:16:40: error: no member named 'hardware_destructive_interference_size' in namespace 'std'
   const size_t cache_line_size = std::hardware_destructive_interference_size;
```

It is recommended not to delete the Android macro restrictions in the prefetch.cpp file.
#if defined(__cpp_lib_hardware_interference_size) && !defined(BOTAN_TARGET_OS_IS_ANDROID)

**Comment by reneme** [comment]
I find it surprising that this doesn't come up in our Android CI build. So basically, some Android environments define `__cpp_lib_hardware_interference_size` but don't provide `std::hardware_destructive_interference_size`?

**Comment by jimwang118** [comment]
> I find it surprising that this doesn't come up in our Android CI build. So basically, some Android environments define `__cpp_lib_hardware_interference_size` but don't provide `std::hardware_destructive_interference_size`?

Yes, __cpp_lib_hardware_interference_size is defined on my test environment, but std::hardware_destructive_interference_size is not supported.
I used Ubuntu to build an Android test environment.

**Comment by reneme** [comment]
That's our setup in CI as well. We use GitHub Action's Ubuntu 22.04 image with the Android NDK R26 from here: https://dl.google.com/android/repository/android-ndk-r26-linux.zip

And that seems to work fine. Note that [with R26 the libc++ was updated](https://github.com/android/ndk/wiki/Changelog-r26#changelog). Perhaps you're using an older version of the NDK? Please also note that upstream [Botan claims support for the latest NDK](https://botan.randombit.net/handbook/support.html) only.

**Comment by jimwang118** [comment]
Yes, the NDK version I use is r25c.

**Comment by reneme** [comment]
I'm not well versed with NDK. But perhaps, If it offers a `#define` which allows you to detect older versions and disable the macro based on that?

Along those lines:

```C++
// Note: Android NDK below 26 is lying about its support for
//       this C++17 feature.
#if defined(__cpp_lib_hardware_interference_size) &&
    (!defined(BOTAN_TARGET_OS_IS_ANDROID) || defined(NDK_IS_NEWER_THAN_25))
```

**Comment by jimwang118** [comment]
> I'm not well versed with NDK. But perhaps, If it offers a `#define` which allows you to detect older versions and disable the macro based on that?
> 
> Along those lines:
> 
> ```c++
> // Note: Android NDK below 26 is lying about its support for
> //       this C++17 feature.
> #if defined(__cpp_lib_hardware_interference_size) &&
>     (!defined(BOTAN_TARGET_OS_IS_ANDROID) || defined(NDK_IS_NEWER_THAN_25))
> ```

Yes, it is possible to add this version judgment.

**Comment by reneme** [comment]
Please feel free to create a pull request with the patch. 🙂

**Comment by randombit** [comment]
For 3.2 you could fix the build simply by guarding this use with a macro again.

However the reason we are requiring the latest NDK is that until NDK 26, the libc++ shipped with the NDK was many years out of date (Google was updating clang in new NDK releases, but *not* libc++). This restricted what we could do in various ways, where something worked with every other compiler we supported but not the NDK.

We're not going to fix this for 3.3 and in fact we'll start using some of the C++ library features which we were previously held back from using due to the NDK 25 libc++.

Note that Google themselves considers NDK < 26 [Unsupported](https://github.com/android/ndk/wiki/Unsupported-Downloads)
